### ISSUE TITLE
Enable cleanup recipes to have both "find" and "fix" flows

### ISSUE THREAD

**Issue by aegershman** [openrewrite/rewrite#282]
Externalizing/parking this thought for later so I don't forget. Feel free to close whenever. This is not an immediately pressing issue in any sense.

In order to demonstrate similarity / feature-compatibility with existing linting tooling which operate on _identifying_ a "violation", some of the `cleanup`-type recipes should be able to both `find` and `remediate` a given problem. E.g., being able to both "find" `UtilityClass` violations (or `FinalLocalVariable` violations, or whatever the case may be) as separate from a "fix" of those violations.

**Comment by sambsnyd** [comment]
Demonstrating similarity or feature parity with linters is not a goal of ours. You can do a dry run of any recipe with the "warn" build plugin task (although the reporting/ux there leaves something to be desired).

**Comment by aegershman** [comment]
Fair; I think what I'm trying to suss out (/articulate?) through dumping thoughts in this issue is-- whether there's value in being able to identify and count "opportunity for refactoring" and "number of times <recipes, visitors> operated on a codebase", for any given recipe or visitor.

So for example, being able to identify how _many_ "SpringRunnerToSpringExtensions" (or whatever) fixes a recipe could/has remediated. If you perform an operation on a file, you'll get a git diff on one file of the changes made, but not "how many of each" recipe/visitor was ran; would it be also valuable to have a count of how many specific locations the recipe/visitor "could" have vs "did" repair? Like, how many total opportunities for a recipe to make a fix are there (vs just the git diff showing the end result)?

And if so, would having a generalized pattern of "find(/identify)" and "remediate" be two separate steps being predicates to value-metrics. Assuming that's valuable at all.

This is not a pressing concern- let alone something to consider anytime soon. I didn't do an amazing job articulating these thoughts here or in the first issue body. I guess I'm just getting the idea out on some sort of "paper".

**Comment by jkschneider** [comment]
@aegershman The git diff is just one of many possible outputs from `Result`. Ultimately, I see us fully implementing `SemanticallyEqual`, and being able to represent both the specifics and the number of semantic differences between a "before" and "after" source file, whether those semantic differences arose from a Recipe execution or not.

One such larger goal of `SemanticallyEqual` is to provide a `git sdiff` plugin that no longer shows whitespace and other semantically equal change (like method re-ordering) as a change.

The bottom line is Recipes need not make any distinction between find and fix.
